src/opm/parser/eclipse/EclipseState/Schedule/Group/GuideRate.cpp

```diff
@@ -0,0 +1,130 @@
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <opm/parser/eclipse/EclipseState/Schedule/Schedule.hpp>
+#include <opm/parser/eclipse/EclipseState/Schedule/Group/GuideRate.hpp>
+
+namespace Opm {
+
+GuideRate::GuideRate(const Schedule& schedule_arg) :
+    schedule(schedule_arg)
+{}
+
+
+double GuideRate::get(const std::string& wgname, size_t report_step) const {
+    const auto& dynamic_state = this->values.at(wgname);
+    return dynamic_state[report_step].value;
+}
+
+
+double GuideRate::update(const std::string& wgname, size_t report_step, double oil_pot, double gas_pot, double wat_pot) {
+    const auto& config = this->schedule.guideRateConfig(report_step);
+
+    if (this->values.count(wgname) == 0) {
+        DynamicState<GuideRateValue> dyn_state(this->schedule.getTimeMap(), GuideRateValue{0,0});
+        this->values.insert( std::make_pair(wgname, dyn_state) );
+    }
+
+    if (config.has_well(wgname))
+        this->well_update(wgname, report_step, oil_pot, gas_pot, wat_pot);
+    else if (config.has_group(wgname)) {
+        this->group_update(wgname, report_step, oil_pot, gas_pot, wat_pot);
+    } else
+        throw std::out_of_range("No such well/group: ");
+
+    return this->get(wgname, report_step);
+}
+
+
+void GuideRate::group_update(const std::string& wgname, size_t report_step, double oil_pot, double gas_pot, double wat_pot) {
+    const auto& config = this->schedule.guideRateConfig(report_step);
+    auto sim_time = this->schedule.getTimeMap().seconds(report_step);
+    auto iter = this->values.find(wgname);
+    auto& dyn_state = iter->second;
+    const auto& group = config.group(wgname);
+
+    // If the FORM mode is used we check if the last computation is recent enough;
+    // then we just return.
+    if (group.target == Group2::GuideRateTarget::FORM) {
+        const auto& prev_value = dyn_state[report_step];
+        auto time_diff = sim_time - prev_value.sim_time;
+        if (config.model().update_delay() > time_diff)
+            return;
+    }
+
+
+    double guide_rate;
+
+    if (group.guide_rate == 0 || group.target == Group2::GuideRateTarget::POTN)
+        guide_rate = this->eval_group_pot();
+
+    if (group.target == Group2::GuideRateTarget::INJV)
+        guide_rate = this->eval_group_resvinj();
+
+    if (group.target == Group2::GuideRateTarget::FORM) {
+        const auto& prev_value = dyn_state[report_step];
+        guide_rate = this->eval(config.model(), prev_value);
+    }
+
+    guide_rate = group.guide_rate;
+    dyn_state.update( report_step, GuideRateValue{sim_time, guide_rate} );
+}
+
+
+void GuideRate::well_update(const std::string& wgname, size_t report_step, double oil_pot, double gas_pot, double wat_pot) {
+    const auto& config = this->schedule.guideRateConfig(report_step);
+    auto sim_time = this->schedule.getTimeMap().seconds(report_step);
+    const auto& well = config.well(wgname);
```

Comment:
The well_update is called every time_step not only per report_step. Therefore I think we need to pass the actual time. (Same for groups)

Reply:
OK - good; then I guess we will need both the report_step and the actual time.